### PR TITLE
Refactor Interpreter

### PR DESCRIPTION
**Pull request by b-scholz**
Clear separation between RAM and Interpreter / Synthesiser functionality.

### CODE REVIEW

File: src/RamValue.h [hunk 1]
```diff
@@ -17,9 +17,8 @@
 #pragma once
 
 #include "BinaryFunctorOps.h"
-#include "RamIndex.h"
 #include "RamNode.h"
-#include "RamRecords.h"
+// #include "RamRecords.h"
```
Comment (mmcgr) [inline src/RamValue.h]: Remove this.

File: src/Interpreter.h [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+
+#include "RamStatement.h"
+
+#include <functional>
+#include <ostream>
+#include <vector>
+
+namespace souffle {
+
+class InterpreterRelation {
```
Comment (mmcgr) [inline src/Interpreter.h]: I don't like having all the classes in this file, but maybe changing this is better left to a later refactoring effort that adds interpreter namespaces and folders for file organisation.

File: src/RamValue.h (hunk 1 above)
Comment (b-scholz) [reply src/RamValue.h]: I will fix this in the next pull-request.

File: src/MaxMatching.cpp [hunk 3]
```diff
@@ -22,7 +22,7 @@ namespace souffle {
  */
 
 /** implementation of bredth first search */
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/MaxMatching.cpp]: I will fix this in the next pull request/this is an original comment.

File: src/Interpreter.h (hunk 2 above)
Comment (b-scholz) [reply src/Interpreter.h]: The idea is to fully separate RAM from Interpreter. The RAM interpreter needs to be rewritten entirely and split up into perhaps several files.

File: src/Interpreter.cpp [hunk 4]
```diff
@@ -1064,4 +1050,51 @@ const QueryExecutionStrategy ScheduledExecution = [](
...
+    stats.arity = rel.getArity();
+    stats.size = rel.size();
+    stats.sample_size = count;
+
+    for (std::size_t i = 0; i < rel.getArity(); i++) {
+        // estimate the cardinality of the columns
+        uint64_t card = 0;
+        if (count > 0) {
+            // based on the observed probability
+            uint64_t cur = columns[i].size();
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/Interpreter.cpp]: This is original code and may change later.

File: src/Interpreter.cpp [hunk 5]
```diff
@@ -1064,4 +1050,51 @@ const QueryExecutionStrategy ScheduledExecution = [](
     return ExecutionSummary({order, runtime});
 };
 
+RelationStats RelationStats::extractFrom(const InterpreterRelation& rel, uint32_t sample_size) {
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/Interpreter.cpp]: This is original code and may change later.

**Pull request merged**